File: typescript/packages/core/src/commands/builtin/generic/split.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { specOf } from '../../spec/builtins.ts'
import { FlagView } from '../../spec/types.ts'
import { stripSlash } from '../../../utils/slash.ts'
import { AsyncLineIterator } from '../../../io/async_line_iterator.ts'
import { IOResult } from '../../../io/types.ts'
import { PathSpec } from '../../../types.ts'
import type { CommandFnResult, CommandOpts } from '../../config.ts'
import { resolveSource } from '../utils/stream.ts'
import { extraOperandError } from '../../spec/usage.ts'
import { CommandName } from '../../spec/types.ts'
import { UsageError } from '../../errors.ts'
import { sizeSuffixes } from '../utils/size_suffix.ts'

const ENC = new TextEncoder()

// GNU split's letter set: every uppercase power letter plus b, and
// lowercase k/m only (pinned against coreutils 9.7). Unlike od, split is
// base-10 only: hex and octal spellings are invalid numbers.
const BYTE_UNITS = sizeSuffixes('bkKmMEGPQRTYZ')
const BYTE_SUFFIXES = Object.keys(BYTE_UNITS).sort((a, b) => b.length - a.length)
const DIGITS = /^\d+$/
const HEX_DIGITS = /^[0-9a-fA-F]+$/
const TRY_HELP = "\nTry 'split --help' for more information."

function parseBytesValue(value: string): number {
  const suffix = BYTE_SUFFIXES.find((u) => value.endsWith(u))
  const digits = suffix === undefined ? value : value.slice(0, -suffix.length)
  if (!DIGITS.test(digits) || Number.parseInt(digits, 10) === 0) {
    throw new UsageError(`split: invalid number of bytes: '${value}'`, 1)
  }
  return Number.parseInt(digits, 10) * (suffix === undefined ? 1 : (BYTE_UNITS[suffix] ?? 1))
}

function parseLinesValue(value: string): number {
  if (!DIGITS.test(value) || Number.parseInt(value, 10) === 0) {
    throw new UsageError(`split: invalid number of lines: '${value}'`, 1)
  }
  return Number.parseInt(value, 10)
}

// GNU quotes only the N of an `l/N` or `K/N` chunk spec in the error.
function parseChunksValue(value: string): number {
  const tail = value.split('/').at(-1) ?? value
  if (!DIGITS.test(tail) || Number.parseInt(tail, 10) === 0) {
    throw new UsageError(`split: invalid number of chunks: '${tail}'`, 1)
  }
  return Number.parseInt(tail, 10)
}

function parseSuffixLength(value: string): number {
  if (!DIGITS.test(value)) {
    throw new UsageError(`split: invalid suffix length: '${value}'`, 1)
  }
  return Number.parseInt(value, 10)
}

function parseSuffixStart(value: string, hexMode: boolean, suffixLen: number): number {
  if (!(hexMode ? HEX_DIGITS : DIGITS).test(value)) {
    const kind = hexMode ? 'hexadecimal' : 'numerical'
    throw new UsageError(`split: '${value}': invalid start value for ${kind} suffix${TRY_HELP}`, 1)
  }
  const start = Number.parseInt(value, hexMode ? 16 : 10)
  if (start.toString(hexMode ? 16 : 10).length > suffixLen) {
    throw new UsageError(
      `split: numerical suffix start value is too large for the suffix length${TRY_HELP}`,
      1,
    )
  }
  return start
}

function alphaSuffix(index: number, length: number): string {
  const chars: string[] = []
  let n = index
  for (let i = 0; i < length; i++) {
    chars.push(String.fromCharCode('a'.charCodeAt(0) + (n % 26)))
    n = Math.floor(n / 26)
  }
  return chars.reverse().join('')
}

function numericSuffix(index: number, length: number): string {
  const s = String(index)
  return s.length >= length ? s : '0'.repeat(length - s.length) + s
}

function hexSuffix(index: number, length: number): string {
  return index.toString(16).padStart(length, '0')
}

function makePathSpec(virtual: string): PathSpec {
  return new PathSpec({
    virtual,
    directory: virtual,
    resourcePath: stripSlash(virtual),
    resolved: true,
  })
}

function outputPath(
  prefix: string,
  suffix: (index: number, length: number) => string,
  index: number,
  start: number,
  length: number,
  additional: string,
): string {
  return prefix + suffix(index + start, length) + additional
}

function joinLines(lines: readonly Uint8Array[]): Uint8Array {
  let total = 0
  for (const l of lines) total += l.byteLength + 1
  const out = new Uint8Array(total)
  let offset = 0
  for (const l of lines) {
    out.set(l, offset)
    offset += l.byteLength
    out[offset] = 0x0a
    offset += 1
  }
  return out
}

async function* recordIterator(
  source: AsyncIterable<Uint8Array>,
  separator: number,
): AsyncIterable<Uint8Array> {
  let pending = new Uint8Array(0)
  for await (const chunk of source) {
    const merged = new Uint8Array(pending.byteLength + chunk.byteLength)
    merged.set(pending)
    merged.set(chunk, pending.byteLength)
    let start = 0
    for (let index = 0; index < merged.byteLength; index++) {
      if (merged[index] === separator) {
        yield merged.slice(start, index)
        start = index + 1
      }
    }
    pending = merged.slice(start)
  }
  if (pending.byteLength > 0) yield pending
}

function joinRecords(records: readonly Uint8Array[], separator: number): Uint8Array {
  if (separator === 0x0a) return joinLines(records)
  let total = records.length
  for (const record of records) total += record.byteLength
  const out = new Uint8Array(total)
  let offset = 0
  for (const record of records) {
    out.set(record, offset)
    offset += record.byteLength
    out[offset] = separator
    offset += 1
  }
  return out
}

export async function splitGeneric(
  paths: PathSpec[],
  opts: CommandOpts,
  stream: (p: PathSpec) => AsyncIterable<Uint8Array>,
  write: (p: PathSpec, data: Uint8Array) => Promise<void>,
): Promise<CommandFnResult> {
  const fl = new FlagView(opts.flags, specOf('split'))
  if (paths.length > 2) throw extraOperandError(CommandName.SPLIT, paths[2]?.rawPath ?? '')
  const prefixPath = paths.length >= 2 && paths[1] !== undefined ? paths[1].mountPath : 'x'
  const linesValue = fl.asStr('lines')
  const bytesValue = fl.asStr('bytes')
  const numberValue = fl.asStr('number')
  const lengthValue = fl.asStr('suffix_length')
  const numericValue = fl.raw('numeric_suffixes')
  const hexValue = fl.raw('hex_suffixes')
  const separatorValue = fl.asStr('separator')
  const linesFlag = typeof linesValue === 'string' ? linesValue : null
  const bFlag = typeof bytesValue === 'string' ? bytesValue : null
  const nFlag = typeof numberValue === 'string' ? numberValue : null
  const aFlag = typeof lengthValue === 'string' ? lengthValue : null
  const dFlag = numericValue !== undefined
  const xFlag = hexValue !== undefined
  const suffixLenRaw = aFlag !== null ? parseSuffixLength(aFlag) : 2
  // GNU reads an explicit `-a 0` as "revert to auto width", which for
  // fixed-width naming is the default length of 2.
  const suffixLen = suffixLenRaw === 0 ? 2 : suffixLenRaw
  const suffixStart =
    typeof numericValue === 'string'
      ? parseSuffixStart(numericValue, false, suffixLen)
      : typeof hexValue === 'string'
        ? parseSuffixStart(hexValue, true, suffixLen)
        : 0
  const additionalSuffix = fl.asStr('additional_suffix') ?? ''
  const separator =
    separatorValue === '\\0'
      ? 0
      : typeof separatorValue === 'string'
        ? (ENC.encode(separatorValue)[0] ?? 0x0a)
        : 0x0a
  const linesPerFile =
    linesFlag !== null ? parseLinesValue(linesFlag) : bFlag === null && nFlag === null ? 1000 : 0
  const byteLimit = bFlag !== null ? parseBytesValue(bFlag) : 0
  const nChunks = nFlag !== null ? parseChunksValue(nFlag) : 0
  const suffixFn = xFlag ? hexSuffix : dFlag ? numericSuffix : alphaSuffix

  let source: AsyncIterable<Uint8Array>
  const first = paths[0]
  if (first !== undefined) {
    source = stream(first)
  } else {
    source = resolveSource(opts.stdin)
  }

  const writes: Record<string, Uint8Array> = {}
  let fileIdx = 0

  if (nChunks > 0) {
    const chunks: Uint8Array[] = []
    let total = 0
    for await (const c of source) {
      chunks.push(c)
      total += c.byteLength
    }
    const allData = new Uint8Array(total)
    let offset = 0
    for (const c of chunks) {
      allData.set(c, offset)
      offset += c.byteLength
    }
    const chunkSize = Math.max(1, Math.ceil(total / nChunks))
    offset = 0
    for (let i = 0; i < nChunks; i++) {
      const part = allData.slice(offset, offset + chunkSize)
      if (part.byteLength === 0) break
      const outPath = outputPath(prefixPath, suffixFn, i, suffixStart, suffixLen, additionalSuffix)
      await write(makePathSpec(outPath), part)
      writes[outPath] = part
      offset += chunkSize
    }
  } else if (byteLimit > 0) {
    let buf = new Uint8Array(0)
    for await (const c of source) {
      const merged = new Uint8Array(buf.byteLength + c.byteLength)
      merged.set(buf, 0)
      merged.set(c, buf.byteLength)
      buf = merged
      while (buf.byteLength >= byteLimit) {
        const outPath = outputPath(
          prefixPath,
          suffixFn,
          fileIdx,
          suffixStart,
          suffixLen,
          additionalSuffix,
        )
        const data = buf.slice(0, byteLimit)
        await write(makePathSpec(outPath), data)
        writes[outPath] = data
        buf = buf.slice(byteLimit)
        fileIdx += 1
      }
    }
    if (buf.byteLength > 0) {
      const outPath = outputPath(
        prefixPath,
        suffixFn,
        fileIdx,
        suffixStart,
        suffixLen,
        additionalSuffix,
      )
      await write(makePathSpec(outPath), buf)
      writes[outPath] = buf
    }
  } else {
    const lineBuf: Uint8Array[] = []
    const iter =
      separator === 0x0a ? new AsyncLineIterator(source) : recordIterator(source, separator)
    for await (const line of iter) {
      lineBuf.push(line)
      if (lineBuf.length >= linesPerFile) {
        const outPath = outputPath(
          prefixPath,
          suffixFn,
          fileIdx,
          suffixStart,
          suffixLen,
          additionalSuffix,
        )
        const data = joinRecords(lineBuf, separator)
        await write(makePathSpec(outPath), data)
        writes[outPath] = data
        lineBuf.length = 0
        fileIdx += 1
      }
    }
    if (lineBuf.length > 0) {
      const outPath = outputPath(
        prefixPath,
        suffixFn,
        fileIdx,
        suffixStart,
        suffixLen,
        additionalSuffix,
      )
      const data = joinRecords(lineBuf, separator)
      await write(makePathSpec(outPath), data)
      writes[outPath] = data
    }
  }
  return [null, new IOResult({ writes })]
}

File: python/mirage/commands/builtin/generic/split.py
from collections.abc import AsyncIterator, Awaitable, Callable, Iterator
from dataclasses import dataclass
from enum import Enum
from functools import partial
from itertools import islice

from mirage.commands.builtin.constants import (SPLIT_BYTE_SUFFIXES,
                                               SPLIT_BYTE_UNITS,
                                               SPLIT_COUNT_PATTERN,
                                               SPLIT_DIGITS, SPLIT_HEX_DIGITS,
                                               SPLIT_TRY_HELP, UINTMAX)
from mirage.commands.builtin.utils.stream import resolve_source
from mirage.commands.errors import UsageError
from mirage.commands.quote import quote_text
from mirage.commands.spec.types import CommandName
from mirage.commands.spec.usage import extra_operand_error
from mirage.io.async_line_iterator import AsyncLineIterator
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec
from mirage.utils.key_prefix import mounted_path


class ChunkKind(Enum):
    """The three ``-n`` modes: byte chunks, record chunks, round robin."""

    BYTES = "bytes"
    LINES = "l"
    ROUND_ROBIN = "r"


_CHUNK_KIND_PREFIXES = (("l/", ChunkKind.LINES), ("r/", ChunkKind.ROUND_ROBIN))


@dataclass(frozen=True, slots=True)
class ChunkSpec:
    """A parsed ``-n`` value.

    Args:
        kind (ChunkKind): how the input is cut.
        count (int): N, the number of chunks.
        only (int | None): K of ``K/N``: the one chunk written to stdout,
            with no output file created at all; None writes every chunk
            to its own file.
    """

    kind: ChunkKind
    count: int
    only: int | None = None


def parse_bytes_value(value: str) -> int:
    """GNU ``split -b`` byte count: base-10 digits plus a size suffix.

    Args:
        value (str): the raw flag value, e.g. ``4``, ``1K``, ``2GiB``.
    """
    suffix = next((u for u in SPLIT_BYTE_SUFFIXES if value.endswith(u)), "")
    digits = value[:-len(suffix)] if suffix else value
    if SPLIT_COUNT_PATTERN.fullmatch(digits) is None or int(digits) == 0:
        raise UsageError(
            f"split: invalid number of bytes: '{quote_text(value)}'", 1)
    return int(digits) * SPLIT_BYTE_UNITS.get(suffix, 1)


def parse_lines_value(value: str) -> int:
    """GNU ``split -l`` line count: base-10 digits, no suffixes.

    Args:
        value (str): the raw flag value.
    """
    if SPLIT_COUNT_PATTERN.fullmatch(value) is None or int(value) == 0:
        raise UsageError(
            f"split: invalid number of lines: '{quote_text(value)}'", 1)
    return int(value)


def parse_chunks_value(value: str) -> ChunkSpec:
    """GNU ``split -n``: ``N``, ``K/N``, ``l/N``, ``l/K/N``, ``r/N``, ``r/K/N``
    (``KIND/K/N`` selects one chunk).

    Args:
        value (str): the raw flag value, e.g. ``4``, ``l/4``, ``2/3``.
    """
    # GNU strips ONE leading `l/` or `r/` and then cuts what is left at
    # its FIRST slash into K and N: a K it cannot parse names the whole
    # remainder, and every other refusal names N, which is the rest of
    # the spec however many slashes that still holds. Measured on
    # coreutils 9.4: `l/xé/4` names `xé/4`, `2/3/4` and `l/2/3/4` name
    # `3/4`, `l//4` names `/4`, `r/l/4` names `l/4`, and `+l/2` and
    # `x/3` name themselves because neither carries a kind prefix.
    # mirage used to name the whole spec for every malformed head and
    # credited that to 9.7; 9.4 disagrees, and so does the accepted set
    # -- a third component is N's problem, not a head component.
    # A K that parses but is 0 or past N is its own refusal, `invalid
    # chunk number`, checked after N (coreutils 9.7: `4/3` and `0/3`
    # name K, `3/0` names N).
    kind = ChunkKind.BYTES
    spec = value
    for prefix, prefixed_kind in _CHUNK_KIND_PREFIXES:
        if value.startswith(prefix):
            kind = prefixed_kind
            spec = value[len(prefix):]
            break
    head, slash, tail = spec.partition("/")
    if slash and SPLIT_COUNT_PATTERN.fullmatch(head) is None:
        raise UsageError(
            f"split: invalid number of chunks: '{quote_text(spec)}'", 1)
    count_raw = tail if slash else head
    if SPLIT_COUNT_PATTERN.fullmatch(count_raw) is None or int(count_raw) == 0:
        raise UsageError(
            f"split: invalid number of chunks: '{quote_text(count_raw)}'", 1)
    count = int(count_raw)
    only: int | None = None
    if slash:
        only = int(head)
        if only == 0 or only > count:
            raise UsageError(
                f"split: invalid chunk number: '{quote_text(head)}'", 1)
    return ChunkSpec(kind, count, only)


def parse_suffix_length(value: str) -> int:
    """GNU ``split -a`` suffix length: base-10 digits, 0 means auto.

    Args:
        value (str): the raw flag value.
    """
    if SPLIT_COUNT_PATTERN.fullmatch(value) is None:
        raise UsageError(
            f"split: invalid suffix length: '{quote_text(value)}'", 1)
    length = int(value)
    # xstrtoumax overflow: past 2**64 - 1 GNU refuses the width at parse
    # time (byte and line counts saturate instead — a count bigger than
    # the input reads the same either way, but a width this size would be
    # built into a file name).
    if length > UINTMAX:
        raise UsageError(
            f"split: invalid suffix length: '{quote_text(value)}': "
            "Value too large for defined data type", 1)
    return length


def parse_suffix_start(value: str, hex_mode: bool, suffix_len: int) -> int:
    """GNU ``--numeric-suffixes=``/``--hex-suffixes=`` start value.

    The refused value is named through gnulib's ``quote()`` like every
    other word split reports, and it comes FIRST in this clause where
    the four count clauses put it last (measured on coreutils 9.4:
    ``split: 'x\\303\\251': invalid start value for numerical
    suffix``).

    Hex digits are lower case only, as GNU's own suffixes are:
    ``--hex-suffixes=A`` is refused (coreutils 9.7).

    Args:
        value (str): the raw start value; hex digits when ``hex_mode``.
        hex_mode (bool): parse base 16 (``--hex-suffixes``) or base 10.
        suffix_len (int): the effective suffix width the start must fit.
    """
    # An empty value (`--numeric-suffixes=`) is a start of 0 that still
    # pins the width, since GNU checks `strspn` over an empty string and
    # keeps the pointer; only an absent value auto-lengthens.
    if value == "":
        return 0
    pattern = SPLIT_HEX_DIGITS if hex_mode else SPLIT_DIGITS
    if pattern.fullmatch(value) is None:
        kind = "hexadecimal" if hex_mode else "numerical"
        raise UsageError(
            f"split: '{quote_text(value)}': invalid start value for "
            f"{kind} suffix" + SPLIT_TRY_HELP, 1)
    start = int(value, 16 if hex_mode else 10)
    if len(format(start, "x" if hex_mode else "d")) > suffix_len:
        raise UsageError(
            "split: numerical suffix start value is too large "
            "for the suffix length" + SPLIT_TRY_HELP, 1)
    return start


def parse_separator(value: str | None) -> bytes:
    """GNU ``split -t`` record separator: exactly one byte, or ``\\0``.

    GNU reads the value as one byte and refuses every other length rather
    than truncating to the first: an empty value is an empty record
    separator and anything longer is a multi-character one, with the
    two-character spelling ``\\0`` carved out as the only way to write a
    NUL on a command line. The length is counted in bytes, so a lone
    non-ASCII character is multi-character too (pinned against coreutils
    9.7). Deliberate divergence, matching truncate: GNU's quotearg escapes
    control characters in the message and mirage quotes the raw value. Not
    covered: GNU also refuses two ``-t`` flags naming different characters,
    which needs a list-valued flag the spec does not have.

    Args:
        value (str | None): the raw flag value, or None when unset.
    """
    if value is None:
        return b"\n"
    if value == "\\0":
        return b"\0"
    encoded = value.encode()
    if not encoded:
        raise UsageError("split: empty record separator", 1)
    if len(encoded) > 1:
        raise UsageError(f"split: multi-character separator '{value}'", 1)
    return encoded


def _chunk_end(index: int, base: int, rem: int) -> int:
    """Byte offset where chunk ``index`` (1-based) ends.

    GNU sizes chunks ``base`` bytes each with the remainder spread one
    byte at a time over the FIRST chunks, so the end is a closed form
    rather than a table of N prefix sums.

    Args:
        index (int): 1-based chunk number.
        base (int): ``size // count``.
        rem (int): ``size % count``.
    """
    return base * index + min(index, rem)


def _byte_chunks(data: bytes, count: int) -> Iterator[bytes]:
    """The byte chunks of ``data`` in order, the way GNU sizes them.

    ``size // count`` bytes each, with the remainder spread one byte at
    a time over the FIRST chunks: 7 bytes in 3 are 3, 2, 2 (coreutils
    9.7). Stops once the input is used up, because every chunk after
    that is empty; ``chunk_parts`` pads and ``chunk_at`` reads past the
    end, so a huge N never costs N slices.

    Args:
        data (bytes): the whole input.
        count (int): N.
    """
    base, rem = divmod(len(data), count)
    pos = 0
    for index in range(count):
        if pos >= len(data):
            return
        size = base + (1 if index < rem else 0)
        yield data[pos:pos + size]
        pos += size


def _line_chunks(data: bytes, count: int, eol: bytes) -> Iterator[bytes]:
    """The line chunks of ``data`` in order, no record cut.

    GNU's ``lines_chunk_split``: the byte boundaries are those of
    ``_byte_chunks`` over ``max(size, count)``, and a chunk runs to the
    first terminator at or after its own last byte, so a record that
    straddles a boundary goes whole to the chunk it started in. A
    record long enough to cover a whole later chunk leaves that chunk
    empty, and a chunk that begins exactly where the previous one ended
    takes the next record. Measured on coreutils 9.7 (``-n l/7`` over
    five 6-byte lines is line1, line2, line3, empty, line4, line5,
    empty). Stops once the input is used up, like ``_byte_chunks``.

    Args:
        data (bytes): the whole input.
        count (int): N.
        eol (bytes): the one-byte record terminator.
    """
    size = max(len(data), count)
    base, rem = divmod(size, count)
    buf = bytearray()
    pos = 0
    chunk = 0
    while pos < len(data) and chunk < count:
        start = max(pos, _chunk_end(chunk + 1, base, rem) - 1)
        found = data.find(eol, start) if start < len(data) else -1
        end, terminated = (found + 1, True) if found >= 0 else (len(data),
                                                                False)
        buf += data[pos:end]
        pos = end
        while terminated or _chunk_end(chunk + 1, base, rem) <= pos:
            if not terminated and pos >= len(data):
                break
            yield bytes(buf)
            buf = bytearray()
            chunk += 1
            if chunk >= count:
                break
            if _chunk_end(chunk + 1, base, rem) > pos:
                terminated = False
    if chunk < count:
        yield bytes(buf)


def _records(data: bytes, eol: bytes) -> list[bytes]:
    """The records of ``data``, a final unterminated one included.

    Args:
        data (bytes): the whole input.
        eol (bytes): the one-byte record terminator.
    """
    records: list[bytes] = []
    pos = 0
    while pos < len(data):
        found = data.find(eol, pos)
        end = found + 1 if found >= 0 else len(data)
        records.append(data[pos:end])
        pos = end
    return records


def _round_robin_chunks(data: bytes, count: int,
                        eol: bytes) -> Iterator[bytes]:
    """The chunks of ``data`` dealt record by record in turn.

    Chunk ``k`` holds every ``count``-th record from the ``k``-th; a
    chunk past the last record is empty, so the walk stops there.

    Args:
        data (bytes): the whole input.
        count (int): N.
        eol (bytes): the one-byte record terminator; a final record
            without one is dealt too.
    """
    records = _records(data, eol)
    for index in range(min(count, len(records))):
        yield b"".join(records[index::count])


def _cut(data: bytes, chunks: ChunkSpec, separator: bytes) -> Iterator[bytes]:
    if chunks.kind is ChunkKind.LINES:
        return _line_chunks(data, chunks.count, separator)
    if chunks.kind is ChunkKind.ROUND_ROBIN:
        return _round_robin_chunks(data, chunks.count, separator)
    return _byte_chunks(data, chunks.count)


def chunk_parts(data: bytes, chunks: ChunkSpec,
                separator: bytes) -> Iterator[bytes]:
    """Every chunk of ``data`` under one ``-n`` spec, in order.

    Exactly N chunks, the empty tail included, one at a time: the
    caller writes each to its file as it arrives, so N files never
    mean N chunks held at once.

    Args:
        data (bytes): the whole input.
        chunks (ChunkSpec): the parsed ``-n`` value.
        separator (bytes): the record terminator (``-t``).
    """
    produced = 0
    for part in _cut(data, chunks, separator):
        yield part
        produced += 1
    for _ in range(chunks.count - produced):
        yield b""


def chunk_at(data: bytes, chunks: ChunkSpec, separator: bytes,
             index: int) -> bytes:
    """Chunk ``index`` (1-based) of ``data`` under one ``-n`` spec.

    ``K/N`` wants one chunk, so only the chunks before it are cut, and
    a K past the input's last byte is empty at no cost per skipped
    chunk (``-n 2/1000000000`` over 8 bytes is ``b``, instant in GNU).

    Args:
        data (bytes): the whole input.
        chunks (ChunkSpec): the parsed ``-n`` value.
        separator (bytes): the record terminator (``-t``).
        index (int): K.
    """
    return next(islice(_cut(data, chunks, separator), index - 1, None), b"")


_ALPHA_SUFFIXES = "abcdefghijklmnopqrstuvwxyz"
_NUMERIC_SUFFIXES = "0123456789"
_HEX_SUFFIXES = "0123456789abcdef"


def _to_base(value: int, alphabet: str, width: int) -> str:
    base = len(alphabet)
    chars: list[str] = []
    for _ in range(width):
        chars.append(alphabet[value % base])
        value //= base
    return "".join(reversed(chars))


def _suffix_name(index: int, alphabet: str, auto: bool, width: int,
                 start: int) -> str:
    """One output-file suffix, GNU next_file_name style.

    With no explicit width and no explicit start value the suffix
    auto-lengthens, reserving the last alphabet character as a prefix —
    aa..yz, then zaaa..zyzz, then zzaaaa.. (00..89 then 9000..9899 then
    990000.. for -d); band k holds (B-1)*B**(k+1) names behind k reserved
    characters. An explicit -a width or a --numeric/hex-suffixes start
    value pins the width, and running past B**width is GNU's exhaustion
    error with the chunks already written kept (pinned against coreutils
    9.7). Deliberate divergence: GNU with a hex start whose leading digit
    is the reserved 'f' (--hex-suffixes=f0) walks past its alphabet and
    names files with non-hex characters; mirage exhausts cleanly.

    Args:
        index (int): zero-based output file ordinal.
        alphabet (str): suffix alphabet (alpha, numeric or hex).
        auto (bool): auto-lengthen instead of erroring at the width.
        width (int): fixed suffix width when ``auto`` is false.
        start (int): first suffix value (numeric/hex start, else 0).
    """
    base = len(alphabet)
    if auto:
        band = 0
        capacity = (base - 1) * base
        while index >= capacity:
            index -= capacity
            band += 1
            capacity *= base
        return alphabet[-1] * band + _to_base(index, alphabet, band + 2)
    value = start + index
    if value >= base**width:
        raise UsageError("split: output file suffixes exhausted", 1)
    return _to_base(value, alphabet, width)


def _out_spec(anchor: PathSpec | None, out_path: str) -> PathSpec:
    """Spec for an output file on the operand's mount.

    Args:
        anchor (PathSpec | None): Operand on the executing mount, read for
            its prefix; None when the only input is stdin.
        out_path (str): The output's mount-local key.
    """
    if anchor is None:
        return PathSpec.from_str_path(out_path)
    return mounted_path(anchor, "/" + out_path.lstrip("/"))


async def split(
    paths: list[PathSpec],
    *,
    read_stream: Callable[..., AsyncIterator[bytes]],
    write_bytes: Callable[..., Awaitable[None]],
    stdin: ByteSource | None = None,
    lines_per_file: int = 0,
    byte_limit: int = 0,
    chunks: ChunkSpec | None = None,
    suffix_len: int = 2,
    suffix_auto: bool = True,
    numeric_suffix: bool = False,
    hex_suffix: bool = False,
    suffix_start: int = 0,
    additional_suffix: str = "",
    separator: bytes = b"\n",
) -> tuple[ByteSource | None, IOResult]:
    if len(paths) > 2:
        raise extra_operand_error(CommandName.SPLIT, paths[2].raw_path
                                  or paths[2].virtual)
    prefix_name = paths[1].mount_path if len(paths) >= 2 else "x"
    anchor = paths[1] if len(paths) >= 2 else paths[0] if paths else None
    if lines_per_file == 0 and byte_limit == 0 and chunks is None:
        lines_per_file = 1000
    suffix_fn = partial(
        _suffix_name,
        alphabet=(_HEX_SUFFIXES if hex_suffix else
                  _NUMERIC_SUFFIXES if numeric_suffix else _ALPHA_SUFFIXES),
        auto=suffix_auto,
        width=suffix_len,
        start=suffix_start)

    if paths:
        source: AsyncIterator[bytes] = read_stream(paths[0])
    else:
        source = resolve_source(stdin)

    writes: dict[str, ByteSource] = {}
    file_idx = 0

    if chunks is not None:
        all_data = b"".join([chunk async for chunk in source])
        if chunks.only is not None:
            # `K/N` writes the one chunk to stdout and no file at all.
            return chunk_at(all_data, chunks, separator,
                            chunks.only), IOResult()
        # Every chunk gets its file, an empty one included: GNU creates
        # N files for `-n N` however short the input is.
        for i, part in enumerate(chunk_parts(all_data, chunks, separator)):
            out_path = (prefix_name + suffix_fn(i) + additional_suffix)
            await write_bytes(_out_spec(anchor, out_path), part)
            writes[out_path] = part
    elif byte_limit > 0:
        buf = bytearray()
        async for chunk in source:
            buf.extend(chunk)
            while len(buf) >= byte_limit:
                out_path = (prefix_name + suffix_fn(file_idx) +
                            additional_suffix)
                data = bytes(buf[:byte_limit])
                await write_bytes(_out_spec(anchor, out_path), data)
                writes[out_path] = data
                buf = buf[byte_limit:]
                file_idx += 1
        if buf:
            out_path = (prefix_name + suffix_fn(file_idx) + additional_suffix)
            data = bytes(buf)
            await write_bytes(_out_spec(anchor, out_path), data)
            writes[out_path] = data
    else:
        line_buf: list[bytes] = []
        if separator == b"\n":
            records: AsyncIterator[bytes] = AsyncLineIterator(source)
        else:
            raw = b"".join([chunk async for chunk in source])
            records = _record_iterator(raw, separator)
        async for line in records:
            line_buf.append(line)
            if len(line_buf) >= lines_per_file:
                out_path = (prefix_name + suffix_fn(file_idx) +
                            additional_suffix)
                data = separator.join(line_buf) + separator
                await write_bytes(_out_spec(anchor, out_path), data)
                writes[out_path] = data
                line_buf = []
                file_idx += 1
        if line_buf:
            out_path = (prefix_name + suffix_fn(file_idx) + additional_suffix)
            data = separator.join(line_buf) + separator
            await write_bytes(_out_spec(anchor, out_path), data)
            writes[out_path] = data

    return None, IOResult(writes=writes)


async def _record_iterator(data: bytes,
                           separator: bytes) -> AsyncIterator[bytes]:
    # Every separator terminates a record, so only the final unterminated
    # remainder is dropped when empty; a second trailing separator still
    # yields the empty record it terminates (GNU).
    records = data.split(separator)
    if records and not records[-1]:
        records.pop()
    for record in records:
        yield record


__all__ = ["ChunkKind", "ChunkSpec", "chunk_parts", "split"]
